File: Stacks/DoubleStack.cpp
#include <bits/stdc++.h>
using namespace std;

template <class T>
class DoubleStack
{
    int n1, n2;
    int capacity;
    T* data;

public:

    DoubleStack(int cap)
    {
        capacity = cap;
        data = new T[capacity];
        n1 = 0;
        n2 = cap-1;
    }

    int getSize1() {return this.n1;}
    int getSize2() {return this->capacity - this.n2 - 1;}

    bool isEmpty1() {return n1 == 0;}
    bool isEmpty2() {return n2 == capacity-1;}
    bool isFully() {return n1 > n2;}

    bool push1(T value)
    {
        if(isFully()) false;
        data[n1++] = value;
        return true;
    }
    bool push2(T value)
    {
        if(isFully()) return false;
        data[n2--] = value;
        return true;
    }
    
    T top1()
    {
        if(isEmpty1()) throw("Stack is Empty.");
        return data[n1-1];
    }
    T top2()
    {
        if(isEmpty2()) throw("Stack is Empty.");
        return data[n2+1];
    }

    T pop1()
    {
        if(isEmpty1()) throw("Stack is Empty.");
        T value = data[n1-1];
        n1--;
        return value;
    }
    T pop2()
    {
        if(isEmpty2()) throw("Stack is Empty.");
        T value = data[n2+1];
        n2++;
        return value;
    }
};

File: Stacks/DoubleStack.java
public class DoubleStack<E> {

    private static final int CAPACITY = 1000;

    private E[] data;
    private int top1 = -1;
    private int top2 = CAPACITY;

    public DoubleStack() {
        this(CAPACITY);
    }

    public DoubleStack(int capacity) {
        data = (E[]) new Object[capacity];
    }

    public void push1(E e) throws IllegalStateException {
        if (size1() + size2() == data.length) {
            throw new IllegalStateException("Stack is full");
        }
        data[++top1] = e;
    }

    public void push2(E e) throws IllegalStateException {
        if (size1() + size2() == data.length) {
            throw new IllegalStateException("Stack is full");
        }
        data[--top2] = e;
    }

    public E pop1() {
        if (isEmpty1()) return null;
        E answer = data[top1];
        data[top1--] = null;
        return answer;
    }

    public E pop2() {
        if (isEmpty2()) return null;
        E answer = data[top2];
        data[top2++] = null;
        return answer;
    }

    public E top1() {
        if (isEmpty1()) return null;
        return data[top1];
    }

    public E top2() {
        if (isEmpty2()) return null;
        return data[top2];
    }

    public int size1() {
        return top1 + 1;
    }

    public int size2() {
        return data.length - top2;
    }

    public boolean isEmpty1() {
        return top1 == -1;
    }

    public boolean isEmpty2() {
        return top2 == data.length;
    }

    private void resize(int capacity) {
        E[] temp = (E[]) new Object[capacity];
        if (size1() >= 0)
            System.arraycopy(data, 0, temp, 0, size1());
        if (size2() >= 0)
            System.arraycopy(data, top2, temp, capacity - size2(), size2());
        data = temp;
    }
}
